### PR TITLE
[WIP] Persistence

### PR DESCRIPTION
**Pull request by Tylertron1998**
Needs documentation, tests, to actually work (heh) and some more functionality yet.

TODO:
- Events: It would be nice to have some events to hook into, possibly for client-side notifications of database reading/writing, errors, etc.
- Implement a provider: one for testing stuff, probably going to be rethinkdb as that's the database I'm most comfortable with.
- Add test cases: don't want Ruby to reee.
- Add functionality to the PersistenceManager: currently it does literally nothing. I would like it to be able to create models for the IProvider of choice, and provide the "interaction layer" between the user and the Provider.
- Add a logging/tracking system for the database, incase the database goes down half way through the write; we wouldn't want to load corrupted data. Better to load an older version.
- Create a working and useful model for the game state to save.

- Providers: These are "translation layers" between the SU Server, and the DB of choice for the user. This allows for flexable database choice, and for anyone to add a database provider, without having to edit internal data handling. The idea is that they take care of the "database" logic.

- SharperGameStateModel: This is a simple enum to indicate the status of the last read/write/delete/edit from the database.

### CODE REVIEW

File: SharperUniverse.Tests/SharperUniverse.Tests.csproj [hunk 1]
```diff
@@ -1,24 +1,18 @@
-<Project Sdk="Microsoft.NET.Sdk">
-
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (RubyNova) [inline SharperUniverse.Tests/SharperUniverse.Tests.csproj]: These two look exactly identical? Any particular reason this is in here?

File: SharperUniverse/Core/BaseSharperSystem.cs [hunk 2]
```diff
@@ -49,6 +49,14 @@ protected BaseSharperSystem(GameRunner game)
         /// <returns></returns>
         public abstract Task CycleUpdateAsync(int deltaMs);
 
+	    public Task RegisterComponentAsync(SharperEntity entity, BaseSharperComponent component)
```
Comment (RubyNova) [inline SharperUniverse/Core/BaseSharperSystem.cs]: Make sure all the tests and the fluent builder are updated for this change, as well as the examples.

File: SharperUniverse/Core/Builder/ComposeBuilder.cs [hunk 3]
```diff
@@ -26,7 +26,8 @@ internal ComposeBuilder(GameRunner game)
         public async Task StartGameAsync()
         {
             ServerLog.LogInfo("Game server now launching...");
-            _game.Server.Start();
+	        _game.PeristenceManager?.Connect();
```
Comment (RubyNova) [inline SharperUniverse/Core/Builder/ComposeBuilder.cs]: `Connect()` should not be on the interface, as this could very well be implementation-specific. Its best to have it as a private piece of logic in the constructor of the persistence manager.

File: SharperUniverse/Core/Builder/GameBuilder.cs [hunk 4]
```diff
@@ -1,5 +1,4 @@
 ﻿using System.Reflection;
-using Serilog;
```
Comment (RubyNova) [inline SharperUniverse/Core/Builder/GameBuilder.cs]: Uhhh why was this removed?

File: SharperUniverse/Core/GameRunner.cs [hunk 5]
```diff
@@ -18,6 +19,7 @@ public class GameRunner
         internal List<SharperEntity> Entities { get; set; }
         internal ISharperServer Server { get; set; }
         internal int DeltaMs { get; set; }
+	    internal PersistenceManager PeristenceManager { get; set; }
```
Comment (RubyNova) [inline SharperUniverse/Core/GameRunner.cs]: Is it all straight in your mind how this will be interacted with during gameplay?

File: SharperUniverse/Persistence/IDatabaseProvider.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
+using System.Collections.Generic;
```
Comment (RubyNova) [inline SharperUniverse/Persistence/IDatabaseProvider.cs]: Make sure this can modify specific components as well as specific entities inside a save state. You should also keep in mind it needs to be able to overwrite existing save IDs.

File: SharperUniverse/Persistence/LiteDbProvider.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using LiteDB;
+using SharperUniverse.Core;
+
+namespace SharperUniverse.Persistence
+{
+	public class LiteDbProvider : IDatabaseProvider
+	{
+		public string ConnectionString { get; set; }
```
Comment (RubyNova) [inline SharperUniverse/Persistence/LiteDbProvider.cs]: This won't ever change - the setter need not exist - make it so the fluent builder assigns this via the constructor instead.

File: SharperUniverse/Persistence/LiteDbProvider.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+			using (var db = new LiteDatabase(ConnectionString))
+			{
+				var componentData = ComponentsToExportables(components);
+				
+				var temp = new Dictionary<SharperEntity, List<IImportable<BaseSharperComponent>>>();
+				
+				foreach (var component in components)
+				{
+					
+					var cast = component as IExportable<IImportable<BaseSharperComponent>, object>;
```
Comment (RubyNova) [inline SharperUniverse/Persistence/LiteDbProvider.cs]: Hard cast pls, also `object`? Any particular reason its `object`?

File: SharperUniverse/Persistence/LiteDbProvider.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+					Data = temp
+				};
+				
+				var coll = db.GetCollection<SharperSaveModel>("saves");
+				
+				return coll.Insert(save);
+			}
+		}
+
+		public async Task<> Load(int index)
```
Comment (RubyNova) [inline SharperUniverse/Persistence/LiteDbProvider.cs]: `Task<>` not `Task`?

File: SharperUniverse/Persistence/LiteDbProvider.cs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+		{
+			using (var db = new LiteDatabase(ConnectionString))
+			{
+				var temp = db.GetCollection<SharperSaveModel>("saves").FindById(index);
+				foreach (KeyValuePair<SharperEntity, List<IImportable<BaseSharperComponent>>> pair in temp.Data)
+				{
+					foreach (var component in pair.Value)
+					{
+						var system = _systems.Find(x => x.GetType().GetGenericArguments().First() == component.SystemType);
+						if (system == null) continue; //TODO throw here
```
Comment (RubyNova) [inline SharperUniverse/Persistence/LiteDbProvider.cs]: Yeah don't forget about this lol

File: SharperUniverse/Persistence/LiteDbProvider.cs [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+			return Task.CompletedTask;
+		}
+
+		public void Modify(int index, List<BaseSharperComponent> components)
+		{
+			using (var db = new LiteDatabase(ConnectionString))
+			{
+
+				var data = ComponentsToExportables(components);
```
Comment (RubyNova) [inline SharperUniverse/Persistence/LiteDbProvider.cs]: Rename this method so it sounds more like a verb - such as `GetExportablesFromComponents`

File: SharperUniverse/Persistence/LiteDbProvider.cs [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+				{
+					Data = data
+				};
+				db.GetCollection<SharperSaveModel>("saves").Update(index, model);
+			}
+		}
+
+		public void Delete(int index)
+		{
+			throw new System.NotImplementedException();
```
Comment (RubyNova) [inline SharperUniverse/Persistence/LiteDbProvider.cs]: Wew lad.

File: SharperUniverse/Persistence/PersistenceManager.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+			_provider = (IDatabaseProvider)Activator.CreateInstance(provider);
+			_provider.ConnectionString = connectionString;
+		}
+
+		public int Save()
+		{
+			foreach(var component)
+		}
+
+		public List<BaseSharperComponent> Load(int saveIndex)
```
Comment (RubyNova) [inline SharperUniverse/Persistence/PersistenceManager.cs]: Perhaps consider making this a generic method that can return a `List<T>` where `T` is the specified component?

File: SharperUniverse.Tests/SharperUniverse.Tests.csproj (hunk 1 above)
Comment (Tylertron1998) [reply SharperUniverse.Tests/SharperUniverse.Tests.csproj]: Hmm, I didn't touch this. Odd. Possible that Rider removed whitespace and git diff is being strange.

File: SharperUniverse/Core/Builder/ComposeBuilder.cs (hunk 3 above)
Comment (Tylertron1998) [reply SharperUniverse/Core/Builder/ComposeBuilder.cs]: Oops, forgot to remove this. Thanks!

File: SharperUniverse/Core/Builder/GameBuilder.cs (hunk 4 above)
Comment (Tylertron1998) [reply SharperUniverse/Core/Builder/GameBuilder.cs]: I have no clue. Perhaps rider auto removed it. Is anything from that namespace used?

File: SharperUniverse/Persistence/LiteDbProvider.cs (hunk 9 above)
Comment (Tylertron1998) [reply SharperUniverse/Persistence/LiteDbProvider.cs]: Oops

File: SharperUniverse/Persistence/LiteDbProvider.cs (hunk 11 above)
Comment (Tylertron1998) [reply SharperUniverse/Persistence/LiteDbProvider.cs]: Yessir!

File: SharperUniverse/Core/Builder/GameBuilder.cs (hunk 4 above)
Comment (RubyNova) [reply SharperUniverse/Core/Builder/GameBuilder.cs]: Yes - at least it should be. Its our logging tool.
